### PR TITLE
[1/kv-regression] Put KV into a dedicated thread pool

### LINKED ISSUE #20772: [Bug] Move GCS internal kv runs in a dedicated thread
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/ray/issues) and found no similar issues.

### Ray Component

Ray Core

### What happened + What you expected to happen

Internal KV right now will compete with other gcs resources which can increase the latency. internal kv request should have higher priority since they are cheaper and important. We should move them into a dedicated thread.

### Versions / Dependencies

ray master

### Reproduction script

When ray run in stress mode, sometimes the worker will fail to start. 

### Anything else

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
After moving internal kv to grpc, there is a regression in actor launching performance. This PR move the work from main thread to a dedicated thread for internal kv to mitigate it.

## Related issue number

Closes #20772 
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by rkooo567 — COMMENTED** [review]
This LGTM. I think the only concern is how it'd behave with `redis_client_->GetPrimaryContext()->RunArgvAsync`.

Is Redis client initialized with the main thread io service? In this case, can we make sure to comment the callback is not running in the same thread as kv manager thread? 

Alternatively, I think we can create a separate Redis client with kv manager's io services. (maybe having 1~4 more Redis clients won't impact # of connection to Redis)

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -246,7 +246,8 @@ RAY_CONFIG(int32_t, maximum_profile_table_rows_count, 10 * 1000)
 /// When getting objects from object store, max number of ids to print in the warning
 /// message.
 RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
-
+/// Number of internal kv thread pool
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:249]:
```suggestion
/// Number of threads for the GCS internal kv service.
```

File: src/ray/gcs/gcs_server/gcs_kv_manager.h [hunk 2]
```diff
@@ -49,8 +49,16 @@ class GcsInternalKVManager : public rpc::InternalKVHandler {
                             rpc::InternalKVKeysReply *reply,
                             rpc::SendReplyCallback send_reply_callback);
 
+  instrumented_io_context &GetEventLoop() { return io_service_; }
+
+  void Stop();
+
  private:
   std::shared_ptr<RedisClient> redis_client_;
+  // The io service used by internal kv
```
Comment (ericl) [inline src/ray/gcs/gcs_server/gcs_kv_manager.h:58]:
```suggestion
  // The io service used by internal kv.
```

Please make sure comments are full sentences.

**Review by ericl — COMMENTED** [review]
> This LGTM. I think the only concern is how it'd behave with redis_client_->GetPrimaryContext()->RunArgvAsync.

Yeah, also wondering how it works with the redis here.

**Comment by iycheng** [comment]
@rkooo567  It's thread-safe ([code](https://sourcegraph.com/github.com/ray-project/ray/-/blob/src/ray/gcs/redis_context.cc?L427)), so I'd prefer not updating it to this way.

I'll add comments about the rest.

**Comment by rkooo567** [comment]
@iycheng btw, one of potential cons of this approach is that the SEND_REPLY_CALLBACK will call `response_writer_.Finish(*reply_, RayStatusToGrpcStatus(status), this);` in the thread it is called (that says, the reply will be written from the main thread). Not sure if this is bad or not.

**Review by rkooo567 — APPROVED** [review]

**Comment by iycheng** [comment]
> @iycheng btw, one of potential cons of this approach (using main thread for redis client) is that the SEND_REPLY_CALLBACK will call `response_writer_.Finish(*reply_, RayStatusToGrpcStatus(status), this);` in the thread it is called (that says, the reply will be written from the main thread). Not sure if this is bad or not.

Good point. I agree with this. The cases here are really bad :( :( :( :( 

I think it's about short-term fixes and long-term fixes. 
The root issue is that callback happens in the place where the class can't control and we need fiber for this. I'll put a shared Redis client inside this class.

File: src/ray/gcs/gcs_server/gcs_kv_manager.cc [hunk 3]
```diff
@@ -17,10 +17,32 @@
...
+    threads_.emplace_back(std::make_unique<std::thread>(cb));
+  }
+  redis_client_ = std::make_unique<RedisClient>(redis_options);
+  redis_client_->Connect(io_service_);
+}
+
+void GcsInternalKVManager::Stop() {
+  io_service_.stop();
+  for (auto &t : threads_) {
+    t->joinable();
```
Comment (mwtian) [inline src/ray/gcs/gcs_server/gcs_kv_manager.cc:36]: RAY_CHECK or remove?

**Review by mwtian — APPROVED** [review]

**Review by ericl — APPROVED** [review]

**Comment by iycheng** [comment]
test failure, investigating

**Comment by iycheng** [comment]
> > @iycheng btw, one of potential cons of this approach (using main thread for redis client) is that the SEND_REPLY_CALLBACK will call `response_writer_.Finish(*reply_, RayStatusToGrpcStatus(status), this);` in the thread it is called (that says, the reply will be written from the main thread). Not sure if this is bad or not.
> 
> Good point. I agree with this. The cases here are really bad :( :( :( :(
> 
> I think it's about short-term fixes and long-term fixes. The root issue is that callback happens in the place where the class can't control and we need fiber for this. I'll put a shared Redis client inside this class.

@rkooo567  Back to this thread, the Redis implementation is using a lock to make it thread-safe, and also because of what you mentioned, new Redis seems to make things better:

```
  actors_per_second = 544.9673515585151
  num_actors = 10000
  time = 18.34972310066223
```

File: src/ray/common/ray_config_def.h [hunk 4]
```diff
@@ -246,7 +246,8 @@ RAY_CONFIG(int32_t, maximum_profile_table_rows_count, 10 * 1000)
 /// When getting objects from object store, max number of ids to print in the warning
 /// message.
 RAY_CONFIG(uint32_t, object_store_get_max_ids_to_print_in_warning, 20)
-
+/// Number of threads for the GCS internal kv service.
+RAY_CONFIG(uint32_t, gcs_internal_kv_thread_num, 4);
```
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:250]: I not suspect 1 might be sufficient to achieve the same performance? Do you mind experimenting it?

**Review by rkooo567 — APPROVED** [review]
One last comment (not a hard requirement)! LGTM. Also it seems like the performance is identical as before in this case?

File: src/ray/common/ray_config_def.h (hunk 4 above)
Comment (iycheng) [reply src/ray/common/ray_config_def.h:250]: I think it should be ok given that in start up, it's actually just 30k kv requests not a big deal actually.

Comment (rkooo567) [reply src/ray/common/ray_config_def.h:250]: oh I feel like you might misinterpret what I said (and I had typo lol). I meant wouldn't having 1 thread enough to achieve the same performance? (or you are saying you prefer to keep it 4?)

Comment (iycheng) [reply src/ray/common/ray_config_def.h:250]: i mean 1 should be enough and I'm running it. (30k request is not a big deal => 1 thread might be enough) :p

Comment (iycheng) [reply src/ray/common/ray_config_def.h:250]: yep, they are similar

**Comment by iycheng** [comment]
failed test is flaky on master

**Pull request merged**